### PR TITLE
Improve call req validation error handling paths

### PR DESCRIPTION
**Pull request by jcorbin**
r @kriskowal @ShanniLi 

Especial feedback to the part where I emit a req error event now, rather than assert.

### CODE REVIEW

File: node/v2/handler.js [hunk 1]
```diff
@@ -630,36 +630,33 @@ TChannelV2Handler.prototype.verifyCallRequestFrame =
...
-    } else if (!req.headers || !req.headers.as) {
-        self.logger.error('Expected "as" header to be set for request', {
-            arg1: req.endpoint,
-            callerName: req.headers && req.headers.cn,
-            remoteName: self.remoteName,
-            serviceName: req.serviceName,
-            socketRemoteAddr: self.connection.socketRemoteAddr
-        });
+    if (!req.headers || !req.headers.as) {
+        if (self.requireAs) {
```
Comment (Raynos) [inline node/v2/handler.js]: The whole point of `requireAs` was to make it a programmer error.

Making an outgoing call without a `cn` header should statically throw rather then give you an error in a callback.

**Comment by jcorbin** [comment]
Understood @raynos, I walked this PR back to throwing errors, and we'll have a "should it through" discussion in another PR.  In this PR now all I do is normalize arg1 vs header validation error handling.

File: node/test/max-call-overhead.js [hunk 2]
```diff
@@ -99,18 +106,20 @@ allocCluster.test('request() with large arg1', {
...
     function onResponse(err, resp, arg2, arg3) {
-        assert.ok(err);
-        assert.equal(err.type, 'tchannel.arg1-over-length-limit');
-        assert.equal(err.message,
-            'arg1 length 16385 is larger than the limit 16384'
-        );
+        assert.fail("shouldn't get a response callback");
 
-        assert.equal(null, resp);
+        // assert.ok(err);
```
Comment (ShanniLi) [inline node/test/max-call-overhead.js]: remove the commented code?

Comment (jcorbin) [inline node/test/max-call-overhead.js]: Maybe, it's here in case we to the TODO above about making this a non-throwing path again.

**Comment by ShanniLi** [comment]
:+1: Like the idea to have more immediate/clear errors.

File: node/test/max-call-overhead.js [hunk 3]
```diff
@@ -81,16 +81,23 @@ allocCluster.test('request() with large arg1', {
...
     subTwo.waitForIdentified({
         host: one.hostPort
     }, function onIdentified(err) {
         assert.ifError(err);
 
-        var arg1 = '';
-        for (var i = 0; i < 16 * 1024 + 1; i++) {
-            arg1 += 'a';
-        }
+        // TODO: O(pinions) -- is this really what we want?
```
Comment (kriskowal) [inline node/test/max-call-overhead.js]: O(bservation): There will be cases of computed arg1, possibly including some HTTP translation layer, in which case it should be an operational error, not a hard stop and die.

File: node/test/max-call-overhead.js [hunk 4]
```diff
@@ -99,18 +106,20 @@ allocCluster.test('request() with large arg1', {
...
-    });
+    }
 
     function onResponse(err, resp, arg2, arg3) {
-        assert.ok(err);
-        assert.equal(err.type, 'tchannel.arg1-over-length-limit');
-        assert.equal(err.message,
-            'arg1 length 16385 is larger than the limit 16384'
-        );
+        assert.fail("shouldn't get a response callback");
```
Comment (kriskowal) [inline node/test/max-call-overhead.js]: lintme

File: node/test/max-call-overhead.js [hunk 5]
```diff
@@ -99,18 +106,20 @@ allocCluster.test('request() with large arg1', {
...
+        // assert.ok(err);
+        // assert.equal(err.type, 'tchannel.arg1-over-length-limit');
+        // assert.equal(err.message,
+        //     'arg1 length 16385 is larger than the limit 16384'
+        // );
 
-        assert.end();
+        // assert.equal(null, resp);
+
+        // assert.end();
```
Comment (kriskowal) [inline node/test/max-call-overhead.js]: I recommend restoring this behavior.

**Comment by jcorbin** [comment]
Although, @Raynos will likely shed a tear, not throwing an error does seem less fragile going forward, as it eliminates a potentially fatal crash for Hyperbahn and maybe even xlate that would be out of its control.

In the case where there has been no callback / error event listener given, it will revert to a thrown error.

File: node/v2/handler.js (hunk 1 above)
Comment (kriskowal) [inline node/v2/handler.js]: Did we not implement a version of this that could be safely employed at Hyperbahn ingress?

File: node/v2/handler.js [hunk 6]
```diff
@@ -630,39 +633,34 @@ TChannelV2Handler.prototype.verifyCallRequestFrame =
...
+                arg1: req.endpoint,
+                remoteName: self.remoteName,
+                serviceName: req.serviceName,
+                socketRemoteAddr: self.connection.socketRemoteAddr
+            });
+        }
     }
 
-    return true;
+    return null;
```
Comment (kriskowal) [inline node/v2/handler.js]: Thank you.

**Comment by kriskowal** [comment]
:ok:

**Comment by Raynos** [comment]
@jcorbin 

You should throw an error. it's impossible to hit this in the relay. We enforce cn & as to exist on incoming request so they should always exist on out request.

Please stick to making programming errors be programming errors. Premature FUD of avoiding an uncaught exception is a waste of time. When we get into a bad state we should shout loudly.

Remember that hyperbahn never crashes on uncaught exceptions.

**Pull request merged**